#1739 #2177 OpenIddict authentication: role section, override scope and role key

Eits-Ian opened the pull request:
## Fixes #2177 
- #2177  

## Proposed Changes
- Added role section
- Variable to define an override for the scope or role descriptor

`"ScopeKey": "oi_scp"` to enable OpenIddict
```json
"RequiredRole": [ "User" ]
```

## Successor
- #1478 

## Closes #1739 
- #1739

Eits-Ian committed:
OpenIddict implementation, role section, override scope and role key

"ScopeKey": "oi_scp" to enable OpenIddict
"RequiredRole": [ "User" ]

raman-m commented:
Hi @Eits-Ian !
Thanks for your interest in Ocelot gateway!

Please do the following:
- Merge [PR 1](https://github.com/Eits-Ian/Ocelot/pull/1)
- Rebase feature branch onto develop
- Resolve all merge conflicts

raman-m committed:
Fix warnings

raman-m committed:
Fix CS0649: Field '_rolesAuthorizer' is never assigned to, and will always have its default value null

raman-m committed:
Fix messages

raman-m reviewed (changes requested):
[4 failed tests](https://app.circleci.com/pipelines/github/ThreeMammals/Ocelot/791/workflows/05a46b7b-7acc-4940-9698-78b8f41a5b4d/jobs/1460?invite=true#step-102-247902_116)!
Could you review & fix them please?

Eits-Ian committed:
Fixed errors

Registered IRolesAuthorizer
Fixed Test should_not_be_valid_if_specified_authentication_provider_isnt_registered()

Eits-Ian commented:
Sorry I did not get a chance to look at this till today, should be fixed, have added you as a collaborator, I hadn't looked at this for a couple of years. Need to add process policies, I am really busy at the moment so will not be able to get to this for a while.

Eits-Ian committed:
Fixed test

Fixed test configuration_is_invalid_with_invalid_authentication_provider()

Eits-Ian commented:
The error is at ClaimsToDownstreamPathTests it is not authenticating, I need to take a look at what is happening but it will need to be later

raman-m commented:
> @Eits-Ian commented [on Aug 24](#issuecomment-1691480538)

Thanks for caring!
I am planning to help you in September by having of some pair programming...
Sorry, now I have a lot of things to do with this project...

raman-m committed:
Refactor `AuthenticationOptions`

raman-m committed:
Fix compiling errors

raman-m commented:
The branch has been rebased onto [release/24.0](https://github.com/ThreeMammals/Ocelot/tree/release/24.0) which has top commit https://github.com/ThreeMammals/Ocelot/commit/59b63eab9c2cd2ab33c5752bbc3bd00a079cd74c !

- Further development is required❕ 
- Fixing failed tests is required❗

raman-m commented:
@Eits-Ian Dear author,
Are you still with Ocelot?
And, what's your name? 😉 

**Will this PR close, is PR related to the following issues?**
- #746 
- #679 
- #1308

Please answer ASAP❗

Eits-Ian commented:
Raman,

I need to go back and look at the code to see what I did but from memory, you could specify the roles/scopes but not use an authorization policy and it worked with openiddict.

I haven't got much time to do much with this at the moment, I will need to get back into it as I must update the api, that use this, but that will not be for a few months.

Best,
Ian.

[e IT Solutions]<http://www.e-itsolutions.co.uk/>
[cid:8d0899cc-d007-4e9b-96c0-9264546c6047]
Ian ‑ e IT s​​​​
Technical Director
[cid:06435eea-36c9-44ca-95ca-2d597167df49]      +44 20 3535 7010
[cid:1a240c67-8ec4-458c-b5be-2f288060708b]      +44 20 3701 9721<https://e-itsolutionsukltd.3cx.co.uk/webclient/#/call?phone=>
***@***.***> ***@***.******@***.***>
[e-itsolutions]<http://eits.dev/>       www.eits<http://www.e-itsolutions.co.uk/>.dev

.NET Development ‑ Recruitment Solutions ‑ Retail Solutions

Confidentiality:
This email and any files transmitted with it are confidential and intended solely for the recipient. If you have received this email in error, use of this information (including
disclosure, copying or distribution) may be unlawful. Please notify e IT soln and delete the message immediately.
Views:
Any views expressed in this message are those of the individual sender, except where the sender specifies and with authority, states them to be the views of e IT soln.
Security:
Internet email is not a 100% secure communications medium. Unless encrypted,email messages passing over the internet are not secure and could be intercepted and read by someone else.
Monitoring:
This message is monitored and archived for security purposes. Further communication will signify your consent to this.
Viruses:

This message has been scanned with multiple virus engines e IT soln cannot guarantee that the content of an email communication or any file attachment is virus free or has not been intercepted or amended as it passes over the internet, it is the receiver's responsibility to ensure that received emails comply with their own internal standards.

e IT soln 13485709 cannot accept responsibility for any loss or damage arising from the use of this email or its attachments.
[cid:008f95f2-ea45-4850-9998-b410cdde3693]       Please consider the environment before printing this email message.
________________________________
From: Raman Maksimchuk ***@***.***>
Sent: Tuesday, April 16, 2024 5:33 PM
To: ThreeMammals/Ocelot ***@***.***>
Cc: Ian - e IT s ***@***.***>; Mention ***@***.***>
Subject: Re: [ThreeMammals/Ocelot] OpenIddict implementation, role section, override scope and role key (#1431)


@Eits-Ian<https://github.com/Eits-Ian> Dear author,
Are you still with Ocelot?
And, what's your name? 😉

Will this PR close, is PR related to the following issues?

  *   #746<https://github.com/ThreeMammals/Ocelot/issues/746>
  *   #679<https://github.com/ThreeMammals/Ocelot/issues/679>
  *   #1308<https://github.com/ThreeMammals/Ocelot/issues/1308>

Please answer ASAP❗

—
Reply to this email directly, view it on GitHub<https://github.com/ThreeMammals/Ocelot/pull/1431#issuecomment-2059499121>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AHVFYBVMSCHDYB56KN4Z7C3Y5VHGJAVCNFSM4XHWRGC2U5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEMBVHE2DSOJRGIYQ>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

raman-m commented:
> I haven't got much time to do much with this at the moment, I will need to get back into it as I must update the api, that use this, **but that will not be for a few months**

It should be a matter of weeks, not months!

Eits-Ian commented:
Raman,

With all due respect I have a 3-year clinic project just about to open which I am a part owner of the clinic, and this is my priority, I'm not coding all day long.

Best,
Ian.

[e IT Solutions]<http://www.e-itsolutions.co.uk/>
[cid:f3427529-7e50-451e-89b9-f4e1736211ac]
Ian ‑ e IT s​​​​
Technical Director
[cid:398b4ab0-d8b1-4df8-94d8-8124dfed0f1b]      +44 20 3535 7010
[cid:0036430a-3393-463c-9aee-8664b525352a]      +44 20 3701 9721<https://e-itsolutionsukltd.3cx.co.uk/webclient/#/call?phone=>
***@***.***> ***@***.******@***.***>
[e-itsolutions]<http://eits.dev/>       www.eits<http://www.e-itsolutions.co.uk/>.dev

.NET Development ‑ Recruitment Solutions ‑ Retail Solutions

Confidentiality:
This email and any files transmitted with it are confidential and intended solely for the recipient. If you have received this email in error, use of this information (including
disclosure, copying or distribution) may be unlawful. Please notify e IT soln and delete the message immediately.
Views:
Any views expressed in this message are those of the individual sender, except where the sender specifies and with authority, states them to be the views of e IT soln.
Security:
Internet email is not a 100% secure communications medium. Unless encrypted,email messages passing over the internet are not secure and could be intercepted and read by someone else.
Monitoring:
This message is monitored and archived for security purposes. Further communication will signify your consent to this.
Viruses:

This message has been scanned with multiple virus engines e IT soln cannot guarantee that the content of an email communication or any file attachment is virus free or has not been intercepted or amended as it passes over the internet, it is the receiver's responsibility to ensure that received emails comply with their own internal standards.

e IT soln 13485709 cannot accept responsibility for any loss or damage arising from the use of this email or its attachments.
[cid:de070fb8-de47-497b-a51f-4867b8d956e3]       Please consider the environment before printing this email message.
________________________________
From: Raman Maksimchuk ***@***.***>
Sent: Friday, April 19, 2024 9:48 AM
To: ThreeMammals/Ocelot ***@***.***>
Cc: Ian - e IT s ***@***.***>; Mention ***@***.***>
Subject: Re: [ThreeMammals/Ocelot] OpenIddict implementation, role section, override scope and role key (#1431)


I haven't got much time to do much with this at the moment, I will need to get back into it as I must update the api, that use this, but that will not be for a few months

It should be a matter of weeks, not months!

—
Reply to this email directly, view it on GitHub<https://github.com/ThreeMammals/Ocelot/pull/1431#issuecomment-2066120167>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AHVFYBSQ3IV5RY56IA5WC33Y6DK6HAVCNFSM4XHWRGC2U5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEMBWGYYTEMBRGY3Q>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

raman-m commented:
@Eits-Ian 
I don't understand you!
Should we close the PR or will you develop it in coming future?

Eits-Ian commented:
Raman,

As I said I do need to do it, but I will not be able to get to it for a few months, I currently end up losing days weeks to the clinic, then I get back to some coding but that is only a few hours at a time. In a few months the clinic will be open, and I will get back to this, I wish I had more time now, but I work 7 day weeks and all hours so there is just no more to give.

Best,
Ian.

[e IT Solutions]<http://www.e-itsolutions.co.uk/>
[cid:633356dd-3a1b-417c-ae70-afba4bee33f8]
Ian ‑ e IT s​​​​
Technical Director
[cid:1c0bad6b-6caf-4168-a62d-b361c7a9976c]      +44 20 3535 7010
[cid:ed9002b3-bb8e-4663-9953-eeec5bd12e97]      +44 20 3701 9721<https://e-itsolutionsukltd.3cx.co.uk/webclient/#/call?phone=>
***@***.***> ***@***.******@***.***>
[e-itsolutions]<http://eits.dev/>       www.eits<http://www.e-itsolutions.co.uk/>.dev

.NET Development ‑ Recruitment Solutions ‑ Retail Solutions

Confidentiality:
This email and any files transmitted with it are confidential and intended solely for the recipient. If you have received this email in error, use of this information (including
disclosure, copying or distribution) may be unlawful. Please notify e IT soln and delete the message immediately.
Views:
Any views expressed in this message are those of the individual sender, except where the sender specifies and with authority, states them to be the views of e IT soln.
Security:
Internet email is not a 100% secure communications medium. Unless encrypted,email messages passing over the internet are not secure and could be intercepted and read by someone else.
Monitoring:
This message is monitored and archived for security purposes. Further communication will signify your consent to this.
Viruses:

This message has been scanned with multiple virus engines e IT soln cannot guarantee that the content of an email communication or any file attachment is virus free or has not been intercepted or amended as it passes over the internet, it is the receiver's responsibility to ensure that received emails comply with their own internal standards.

e IT soln 13485709 cannot accept responsibility for any loss or damage arising from the use of this email or its attachments.
[cid:0a911ee5-49f2-4a4e-a1b2-36f2d932832a]       Please consider the environment before printing this email message.
________________________________
From: Raman Maksimchuk ***@***.***>
Sent: Friday, April 19, 2024 2:21 PM
To: ThreeMammals/Ocelot ***@***.***>
Cc: Ian - e IT s ***@***.***>; Mention ***@***.***>
Subject: Re: [ThreeMammals/Ocelot] OpenIddict implementation, role section, override scope and role key (#1431)


@Eits-Ian<https://github.com/Eits-Ian>
I don't understand you!
Should we close the PR or will you develop it in coming future?

—
Reply to this email directly, view it on GitHub<https://github.com/ThreeMammals/Ocelot/pull/1431#issuecomment-2066573390>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AHVFYBSWKG2SK46S7YEZ66LY6EK4RAVCNFSM4XHWRGC2U5DIOJSWCZC7NNSXTN2JONZXKZKDN5WW2ZLOOQ5TEMBWGY2TOMZTHEYA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

raman-m commented:
🆗 Got it! You are super busy… In this case, let me decide what to do with this PR…

raman-m commented on src/Ocelot/Authorization/Middleware/AuthorizationMiddleware.cs:
```diff
@@ -54,6 +60,33 @@ public async Task Invoke(HttpContext httpContext)
                 }
             }
 
+            if (!IsOptionsHttpMethod(httpContext) && IsAuthenticatedRoute(downstreamRoute))
+            {
```
Why did you copy whole `if`-block above?

raman-m replied:
I strongly disapprove of any form of copying❗

raman-m committed:
Quick code review from @raman-m

raman-m committed:
Fix unit tests

raman-m commented:
So, @Eits-Ian … The current implication is failing of 5 acceptance tests❗ See [the build](https://circleci.com/gh/ThreeMammals/Ocelot/3964) results!

raman-m commented:
### Review of acceptance tests required; development to proceed❗

raman-m commented:
## Successor
- #1478 

This dependency can be addressed by giving high priority to merging this PR first, followed by PR #1478.

raman-m commented:
We should introduce a new "scopes matching strategy" feature directly in this PR due to the previously discussed issues in:
- #1739
- #1520

raman-m commented:
### Further Development Required❗
The [5 failed acceptance tests](https://app.circleci.com/pipelines/github/ThreeMammals/Ocelot/2554/workflows/abc07904-c24b-4f76-8781-199f1466b2d7/jobs/4525) need to be addressed!
![image](https://github.com/user-attachments/assets/35e0be82-b49d-4db6-9515-056294875d2d)

raman-m committed:
Merge 'ThreeMammals:develop' into 'Eits-Ian:OpenIddict'

raman-m committed:
Fix unit tests
